h2/src/main/org/h2/mvstore/tx/TransactionMap.java

```diff
@@ -355,16 +372,19 @@ private V set(Object key, TxDecisionMaker<K,V> decisionMaker) {
                 return res;
             }
             decisionMaker.reset();
+            if (timeoutMillis < -1) {
+                return null;
+            }
             if (mapName == null) {
                 mapName = map.getName();
             }
-        } while (transaction.waitFor(blockingTransaction, mapName, key));
+        } while (timeoutMillis != 0 && transaction.waitFor(blockingTransaction, mapName, key, timeoutMillis));
```

Comment:
timeoutMillis != -1

Reply:
0 is used for `NOWAIT`, with `NOWAIT` we shouldn't wait for anything and loop needs to be aborted.

-1 is used for plain `FOR UPDATE` and `Transaction.waitFor()` uses its own default timeout (taken from the session).